### PR TITLE
ABI `verify_proof` with Plonk VerifierData

### LINKED ISSUE #247: Change `verify_proof` signature to accept `VerifierData` instead of VK and Public Input Positions

### PR DESCRIPTION
**Pull request by vlopes11**
Resolves #247

The verifier key was upgraded to contain also the public input positions
to avoid unnecessary complexity while building the static data of the
used circuits.

This change is transparent to the users of plonk library, so we needed
to remove the public inputs positions as a parameter of the ABI and
treat the verifier agnostically as a vec of u8.

### CODE REVIEW

File: rusk-abi/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+## [0.3.0] - 2021-03-16
+
+### Changed
+
+- Updated `verify_proof` to accept verifier data [#247]
+
```
Comment (CPerezz) [inline rusk-abi/CHANGELOG.md:15]: Shouldn't we do a pre-release for that? Or leave this on `Unreleased` for now?

File: rusk-abi/Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "rusk-abi"
-version = "0.2.0"
+version = "0.3.0"
```
Comment (CPerezz) [inline rusk-abi/Cargo.toml:3]: Ditto

**Review by CPerezz — APPROVED** [review]
Some nits. But LGTM

File: rusk-abi/CHANGELOG.md (hunk 1 above)
Comment (ZER0) [reply rusk-abi/CHANGELOG.md:15]: This is correct.
@vlopes11, create from `master` a branch called `rusk-abi-0.3`, as we did for `0.2`, and make sure this PR points to it instead of `master`.
The `Changed` section should be therefore under `Unreleased` as @CPerezz said, and no additional version are added yet.

Plus, the comment should use present tense (see: https://chris.beams.io/posts/git-commit/), and whenever is possible use the verbs of the CHANGELOG's section (So: Add, Remove, Change, Fix). In some cases we will add some aliases so "Update" or "Replace" will be fine too, however in this case it should be better something like:
```
- Change `verify_proof` to take `VerifierData` instead of `VerifierKey` and PI positions
```

File: rusk-abi/Cargo.toml (hunk 2 above)
Comment (ZER0) [reply rusk-abi/Cargo.toml:3]: No version change here for the time being, since we're going to merge into `rusk-abi-0.3`

File: rusk-abi/src/lib.rs [hunk 3]
```diff
@@ -22,6 +22,12 @@ use dusk_abi::{ContractId, Module};
 use dusk_bls12_381::BlsScalar;
 use dusk_jubjub::{JubJubAffine, JubJubScalar};
 
+/// Transcript label initializations used for proof verifications of
+/// RuskModule
+// This is set outside RuskModule to avoid the need to declare the generic type
+// of store to create a proof
+pub const TRANSCRIPT_LABEL: &'static [u8] = b"dusk-network";
```
Comment (ZER0) [inline rusk-abi/src/lib.rs:29]: I don't want that `rusk-abi` become a dependency of circuits or other things like that, so this shouldn't be exposed at all currently.

If the transcript label needs to be taken from somewhere, is not from here.
We might want to create a new issue / PR where we update `rusk-profile` to store and retrieve the transcript label as well.

At the moment I would propose the following solution:
1. Since it's used only on the `verify_proof`, just revert the change. If you want, you can have as a local constant, not exposed, in the `host` file.

2. Make it as "settings" for the host environment, as it is now the Public Parameters. Basically:
```rust
pub struct RuskModule<S> {
    store: S,

    #[cfg(not(target_arch = "wasm32"))]
    pp: &'static dusk_plonk::prelude::PublicParameters,
    #[cfg(not(target_arch = "wasm32"))]
    transcript: &'static [u8],
}
```

Or something like that, since we want to have the exact same flexibility we have with the public parameters.

File: rusk-abi/src/lib.rs [hunk 4]
```diff
@@ -57,6 +63,24 @@ pub enum PublicInput {
...
+        Self::JubJubScalar(s)
+    }
+}
+
+impl From<JubJubAffine> for PublicInput {
+    fn from(p: JubJubAffine) -> PublicInput {
+        Self::Point(p)
+    }
+}
+
```
Comment (ZER0) [inline rusk-abi/src/lib.rs:83]: If you're going to add this, maybe it would be better having a separate module for `PublicInput`.

File: rusk-abi/tests/lib.rs [hunk 5]
```diff
@@ -145,40 +215,61 @@ fn verify_proof() {
     let mut network = NetworkState::<MS>::default();
     network.register_host_module(rusk_mod);
 
-    let contract_id = network.deploy(contract).unwrap();
+    let contract = network.deploy(contract).unwrap();
```
Comment (ZER0) [inline rusk-abi/tests/lib.rs:218]: Revert this change, `network.deploy()` returns a `ContractId` this can be misleading.

File: rusk-abi/tests/lib.rs [hunk 6]
```diff
@@ -129,8 +129,78 @@ fn schnorr_signature() {
...
+    fn padded_circuit_size(&self) -> usize {
+        1 << 3
+    }
+}
+
 #[test]
 fn verify_proof() {
+    let mut circuit = TestCircuit::new(1, 2);
+
+    let label = TRANSCRIPT_LABEL;
```
Comment (ZER0) [inline rusk-abi/tests/lib.rs:183]: As mentioned, this is wrong: rusk-abi shouldn't be the "source" of the transcript label, it's actually the opposite. The `rusk-abi` module should be feed with the transcript label (this, or we just hardcoded).

**Review by ZER0 — CHANGES_REQUESTED** [review]
LGTM but there are few changes.
Also, about the commits:
1. Since we're using a workspace, the commit message should start with the member's name as prefix. So, in your case, you should have something like:
```
rusk-abi:  Update `dusk-plonk` from `0.6` to `0.7-pre`
rusk-abi: Change `verify_proof` to take `VerifierData` 

The verifier key was upgraded to contain also the public input positions
to avoid unnecessary complexity while building the static data of the
used circuits.

This change is transparent to the users of plonk library, so we needed
to remove the public inputs positions as a parameter of the ABI and
treat the verifier agnostically as a vec of u8.

Resolves: #247
```

Also, notice the `Resolves:` at the end of the commit message and not at the beginning. As we said, we're following https://chris.beams.io/posts/git-commit/ as guidelines.

File: rusk-abi/tests/lib.rs (hunk 6 above)
Comment (vlopes11) [reply rusk-abi/tests/lib.rs:183]: But its not clear in that case who is going to feed `rusk-abi` with that

File: rusk-abi/tests/lib.rs (hunk 5 above)
Comment (vlopes11) [reply rusk-abi/tests/lib.rs:218]: A contract is never used if not deployed to a network state so it can benefit from the storage implementations. Shadowing the original declaration (that is never going to be used) with a shorter name seems more meaningful since the only way to access a deployed contract is through its ID

File: rusk-abi/src/lib.rs (hunk 4 above)
Comment (vlopes11) [reply rusk-abi/src/lib.rs:83]: Do you think its worthy to create a module for something that is ephemeral and should last only until the next release of plonk?

File: rusk-abi/src/lib.rs (hunk 3 above)
Comment (vlopes11) [reply rusk-abi/src/lib.rs:29]: I really don't like this label definition repetition around the code base. This can lead to all sort of silly errors when someone simply uses a different transcript initialization

**Comment by vlopes11** [comment]
N/A, changing target branch

File: rusk-abi/src/lib.rs (hunk 3 above)
Comment (CPerezz) [reply rusk-abi/src/lib.rs:29]: We should simply make that impossible by hardcoding this constant in our circuit definitions/interactions

File: rusk-abi/src/lib.rs (hunk 4 above)
Comment (ZER0) [reply rusk-abi/src/lib.rs:80]: Definitely, yes. I don't want the main lib.rs pollute with temporary things, I prefer to have it isolated if the code grows.

To be clear, I agreed with you that we have to implement the `From` traits, so that from the contracts point of view the changes are minimal when we do the proper thing.

However, since this include more temporary code, I want to be sure we isolate it properly.

**Review by ZER0 — CHANGES_REQUESTED** [review]
Thanks for reopened this, Victor.

My previous final review comment seems still applied, see: https://github.com/dusk-network/rusk/pull/248#pullrequestreview-614137387

Please address it.

I also replied to the other comments that still need to be addressed, in a way or another.

**Review by ZER0 — APPROVED** [review]
